firmware/lib/Gantry/Axis.cpp

```diff
@@ -1,311 +1,569 @@
+                    CHANGE);
 }
 
-void reset_axis(AxisConfig *axis_config, Axis *axis)
+/**
+ * @brief Initializes all components of an axis
+ * 
+ * Sets pin modes, configures and enables interrupts and resets to a default state.
+ * 
+ * @param axis Pointer to the Axis to initialize
+ * @param io   Pointer to an AxisIO struct with the I/O specifications
+ *             for this axis
+ */
+static void setup_axis(Axis *axis, const AxisIO *io)
 {
-    setup_struct(axis_config, axis);
+    axis->io = (*io);
+
+    setup_pins(axis);
+    setup_interrupts(axis);
+
+    // reset in case encoder was accidentally triggered by noise on initialization
+    reset_axis(axis);
 }
 
-void setup_axis(AxisConfig *axis_config, Axis *axis)
-{  
-    setup_pins(axis_config);
-    setup_struct(axis_config, axis);
-    setup_interrupts(axis_config, axis);
-    
-    // reseting axis in case encoder has be triggered or a limit switch due to noise on initilization
-    DEBUG_PRINT_VAL("Reseting axis", 0);
-    reset_axis(axis_config, axis);
+/**
+ * @brief Request an axis to start executing the provided motion
+ * 
+ * The combination of the current axis state and the provided motion must all be valid
+ * for the axis to start moving. If anything is invalid an error code will be returned.
+ * 
+ * @param axis   Pointer to the axis to start moving
+ * @param motion Pointer to an AxisMotion struct specifying the motion to execute
+ * 
+ * @return The result of the request to start moving (either AXIS_OK or an appropriate
+ *         error code)
+ */
+static AxisResult start_axis(Axis *axis, AxisMotion *motion)
+{
+    // Reject if the counts don't match the dir
+    if (motion->dir == DIR_POSITIVE &&
+        (motion->counts_accel < 0 || motion->counts_hold < 0 || motion->counts_decel < 0)) {
+           return AXIS_ERR_DIR_MISMATCH;
+    }
+    if (motion->dir == DIR_NEGATIVE &&
+        (motion->counts_accel > 0 || motion->counts_hold > 0 || motion->counts_decel > 0)) {
+           return AXIS_ERR_DIR_MISMATCH;
+    }
+
+    uint32_t total_counts_abs = abs(motion->counts_accel
+                                    + motion->counts_hold
+                                    + motion->counts_decel);
+    // Reject a zero-distance motion
+    if (total_counts_abs == 0) return AXIS_ERR_ZERO_DIST;
+
+    // Reject if we're already moving - must call stop_axis first
+    if (axis->state.moving) return AXIS_ERR_ALREADY_MOVING;
+
+    // Reject if we've hit the far limit switch and are trying to move forward
+    if (motion->dir == DIR_POSITIVE && axis->state.ls_far_pressed) return AXIS_ERR_LS_FAR;
+    // Reject if we've hit the home limit switch and are trying to move backward
+    if (motion->dir == DIR_NEGATIVE && axis->state.ls_home_pressed) return AXIS_ERR_LS_HOME;
+
+    // Configure state
+    axis->state.moving = true;
+    axis->state.velocity = motion->vel_start;
+    axis->state.next_velocity = axis->state.velocity;
```

Comment:
Next velocity is so that if the velocity hasn't changed (i.e. during the holding segment) we can notice that and not bother resetting the timer in the velocity ISR